### ISSUE TITLE
Record syntax within JS objects gets parsed as JS objects

### ISSUE THREAD

**Issue by tsnobip** [rescript-lang/syntax#174]
Here's a quick repro case:

```rescript
type foo = {foo: string}

let bar: {"bar": foo} = {
  "bar": {
    foo: "hello",
  },
}
```
raises:
```
Type Errors
[E] Line 3, column 24:

This has type:
    < bar : < foo : string > Js.t > Js.t
  But somewhere wanted:
    < bar : foo > Js.t
  Types for method bar are incompatible
```

https://rescript-lang.org/try?code=C4TwDgpgBAZg9nKBeKBvecBcUDOwBOAlgHYDmAvgFCUA2EwUARgIb7aoBEL+H2G5yNJShQurXkJEiM2DgAsINGnA4AaYVHLqqQA

Originally filed in rescript-lang/rescript-compiler#4253, the same error happens in reason syntax (facebook/reason#2554)

**Comment by bloodyowl** [comment]
I hit this bug just today 😄

I think it might be good to take the early ReScript days as an opportunity for this breaking change

**Comment by IwanKaramazow** [comment]
Quoting: https://github.com/rescript-lang/rescript-compiler/issues/4253#issuecomment-602978260
> js object syntax desugars to bs.obj ppx, which interprets nested records as just syntax for objects too, since back then it wouldn't have made sense to embed a record into a js object since record compiled to a private representation (array).
> 
> Changing this back would likely break some code? Though it does make sense to change this back now that records compile to objects

The "fix" would be to parse/transform:
```rescript
let x = {
  "bar": {
    foo: "hello",
  },
}
```
as 
```rescript
let x = {
  let __record = {foo: "hello"} 
  {
    "bar": __record
   }
}
```
The question is, what would this break?

**Comment by jfrolich** [comment]
it would only break if you use `bs.obj` and expect it to apply deeply (which I think would almost never happen in real code bases).

**Comment by bloodyowl** [comment]
@IwanKaramazow That looks a bit dangerous (if you want to rely on execution order). 

I'd say the change should probably be made on the PPX, or maybe the new syntax should "replace" the PPX with the expected behavior if possible

**Comment by bobzhang** [comment]
closed since this is already fixed in 9.0.1
